pkg/inventory/inventory-client.go

```diff
@@ -165,20 +226,147 @@ func (cic *ClusterInventoryClient) retrievePreviousInventoryObjects(current *obj
 	if err != nil {
 		return nil, err
 	}
-	cic.pastInventoryObjects = retrievedInventoryInfos
-	cic.retrievedInventoryObjects = true
-	klog.V(4).Infof("prune %d inventory objects found", len(cic.pastInventoryObjects))
-	return retrievedInventoryInfos, nil
+	var clusterInv *resource.Info
+	if len(retrievedInventoryInfos) == 1 {
+		clusterInv = retrievedInventoryInfos[0]
+	} else if len(retrievedInventoryInfos) > 1 {
+		clusterInv, err = cic.mergeClusterInventory(retrievedInventoryInfos)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return clusterInv, nil
+}
+
+// mergeClusterInventory merges the inventory of multiple inventory objects
+// into one inventory object, and applies it. Deletes the remaining unnecessary
+// inventory objects. There should be only one inventory object stored in the
+// cluster after this function. This special case should be very rare.
+//
+// TODO(seans3): Remove this code once we're certain no customers have multiple
+// inventory objects in their clusters.
+func (cic *ClusterInventoryClient) mergeClusterInventory(invInfos []*resource.Info) (*resource.Info, error) {
+	if len(invInfos) == 0 {
+		return nil, nil
+	}
+	// Make the selection of the retained inventory info deterministic.
+	sort.Sort(ordering.SortableInfos(invInfos))
+	retained := invInfos[0]
+	if cic.dryRun {
+		return retained, nil
+	}
+	wrapRetained := WrapInventoryObj(retained)
+	retainedObjs, err := wrapRetained.Load()
+	if err != nil {
+		return nil, err
+	}
+	for i := 1; i < len(invInfos); i++ {
+		merge := invInfos[i]
+		wrapMerge := WrapInventoryObj(merge)
+		mergeObjs, err := wrapMerge.Load()
+		if err != nil {
+			return nil, err
+		}
+		if err := cic.DeleteInventoryObj(merge); err != nil {
+			return nil, err
+		}
+		retainedObjs = object.Union(retainedObjs, mergeObjs)
+	}
+	if err := wrapRetained.Store(retainedObjs); err != nil {
+		return nil, err
+	}
+	retainInfo, err := wrapRetained.GetObject()
+	if err != nil {
+		return nil, err
+	}
+	if err := cic.applyInventoryObj(retainInfo); err != nil {
```

Comment:
What if this step fails ? We have deleted all the past inventory objects in the cluster at this point.

Reply:
Good catch. The apply must happen before the deletes. Changed and added unit tests.